pygam/pygam.py

```diff
@@ -1781,6 +1781,231 @@ def gridsearch(self, X, y, return_scores=False, keep_best=True,
+        """
+        if not self._is_fitted:
+            raise AttributeError('GAM has not been fitted. Call fit first.')
+        if n_bootstraps < 1:
+            raise ValueError('n_bootstraps must be >= 1;'
+                             ' got {}'.format(n_bootstraps))
+        if n_draws < 1:
+            raise ValueError('n_draws must be >= 1;'
+                             ' got {}'.format(n_draws))
+
+        coef_bootstraps, cov_bootstraps = (
+            self._bootstrap_samples_of_smoothing(X, y, weights, n_bootstraps))
+        coef_draws = self._simulate_coef_from_bootstraps(
+            n_draws, coef_bootstraps, cov_bootstraps)
+
+        return coef_draws
+
+    def _bootstrap_samples_of_smoothing(self, X, y, weights, n_bootstraps):
+        """Sample the smoothing parameters using simulated response data."""
+        mu = self.predict_mu(X)  # Wood pg. 198 step 1
+        coef_bootstraps = [self.coef_]
+        cov_bootstraps = [self.statistics_['cov']]
+
+        for _ in range(n_bootstraps - 1):  # Wood pg. 198 step 2
+            # generate response data from fitted model (Wood pg. 198 step 3)
+            y_bootstrap = self.distribution.sample(mu)
+
+            # fit smoothing parameters on the bootstrap data
+            # (Wood pg. 198 step 4)
+            gam = deepcopy(self)
+            gam.set_params(self.get_params())
+            gam.gridsearch(X, y_bootstrap)  # TODO: add weights
+            lam = gam.lam
+
+            # fit coefficients on the original data given the smoothing params
+            # (Wood pg. 199 step 5)
+            gam = deepcopy(self)
+            gam.set_params(self.get_params())
+            gam.lam = lam
+            gam.fit(X, y)  # TODO: add weights
+
+            coef_bootstraps.append(gam.coef_)
+            cov_bootstraps.append(gam.statistics_['cov'])
+        return coef_bootstraps, cov_bootstraps
+
+    def _simulate_coef_from_bootstraps(
+            self, n_draws, coef_bootstraps, cov_bootstraps):
+        """Simulate coefficients using bootstrap samples."""
+        # Sample indices uniformly from {0, ..., n_bootstraps - 1}
+        # (Wood pg. 199 step 6)
+        random_bootstrap_indices = np.random.choice(
+            np.arange(len(coef_bootstraps)), size=n_draws, replace=True)
+
+        # Simulate `n_draws` many random coefficient vectors from a
+        # multivariate normal distribution with mean and covariance given by
+        # the bootstrap samples (indexed by `random_bootstrap_indices`) of
+        # `coef_bootstraps` and `cov_bootstraps`. Because it's faster to draw
+        # many samples from a certain distribution all at once, we make a dict
+        # mapping bootstrap indices to draw indices and use the `size`
+        # parameter of `np.random.multivariate_normal` to sample the draws
+        # needed from that bootstrap sample all at once.
+        bootstrap_index_to_draw_indices = defaultdict(list)
```

Comment:
The naive way to draw samples here would be:
```
coef_draws = np.empty((n_draws, len(self.coef_)))
for i, index_bootstrap in enumerate(random_indices_of_bootstraps):
    coef_draws[i] = np.random.multivariate_normal(
        coef_bootstraps[index_bootstrap],
        cov_bootstraps[index_bootstrap])
```

This fancy footwork using the dictionary `bootstrap_index_to_draw_indices` speeds things up by calling `np.random.multivariate_normal` only once for each bootstrap sample that gets used and by using the  `size`  parameter of `np.random.multivariate_normal` to draw all the samples it will ever need from that particular mean and covariance. Then the `[[draw_indices]]` puts those samples into the right rows. 🎉  💥